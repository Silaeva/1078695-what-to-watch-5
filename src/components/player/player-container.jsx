import React, {useState, useEffect, createRef, useCallback} from "react";
import PropTypes from "prop-types";
import Player from "./player";

const PlayerContainer = (props) => {
  const {currentFilmId} = props;

  const [isPlaying, setIsPlaying] = useState(true);
  const [duration, setDuration] = useState(0);
  const [progress, setProgress] = useState(0);

  const videoRef = createRef();

  useEffect(() => {
    if (videoRef.current) {
      videoRef.current.oncanplay = () => setDuration(Math.floor(videoRef.current.duration));
      videoRef.current.ontimeupdate = () => setProgress(Math.floor(videoRef.current.currentTime));

      if (isPlaying) {
        videoRef.current.play();
      } else {
        videoRef.current.pause();
      }
    }

    return () => {
      if (videoRef.current) {
        videoRef.current.oncanplay = null;
        videoRef.current.ontimeupdate = null;
      }
    };
  }, [isPlaying, videoRef]);

  const onPlayBtnClick = useCallback(() => {
    setIsPlaying(!isPlaying);
  }, [isPlaying]);

  const onFullscreenClick = useCallback(() => {
    videoRef.current.requestFullscreen();
  });

  return (
    <Player
      duration={duration}
      progress={progress}
      onPlayBtnClick={onPlayBtnClick}
      onFullscreenClick={onFullscreenClick}
      renderPlayer={(film) => {
        return (
          <video ref={videoRef} src={film.srcVideo} className="player__video" poster={film.previewImage}></video>
        );
      }}
      isPlaying={isPlaying}
      currentFilmId={currentFilmId} />
  );
};

PlayerContainer.propTypes = {
  currentFilmId: PropTypes.number.isRequired,
};

export default PlayerContainer;
