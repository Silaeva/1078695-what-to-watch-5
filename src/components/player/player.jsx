import React, {useState, useEffect, createRef, useCallback} from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import {getElapsedTime} from "../../utils";
import filmProp from "../../film-prop";

const Player = (props) => {
  const {films, promoFilm, currentFilmId} = props;

  const [isPlaying, setIsPlaying] = useState(true);
  const [duration, setDuration] = useState(0);
  const [progress, setProgress] = useState(0);

  const videoRef = createRef();

  useEffect(() => {
    if (videoRef.current) {
      videoRef.current.oncanplay = () => setDuration(Math.floor(videoRef.current.duration));
      videoRef.current.ontimeupdate = () => setProgress(Math.floor(videoRef.current.currentTime));

      if (isPlaying) {
        videoRef.current.play();
      } else {
        videoRef.current.pause();
      }
    }

    return () => {
      if (videoRef.current) {
        videoRef.current.oncanplay = null;
        videoRef.current.ontimeupdate = null;
      }
    };
  }, [isPlaying, videoRef]);

  const onPlayBtnClick = useCallback(() => {
    setIsPlaying(!isPlaying);
  }, [isPlaying]);

  const onFullscreenClick = useCallback(() => {
    videoRef.current.requestFullscreen();
  });

  const getCurrentFilm = () => {
    if (currentFilmId === promoFilm.id) {
      return promoFilm;
    }
    return films.find((film) => film.id === currentFilmId);
  };

  const currentFilm = getCurrentFilm();

  const togglerPosition = progress / duration * 100;

  return (
    <div className="player">

      <video ref={videoRef} src={currentFilm.srcVideo} className="player__video" poster={currentFilm.previewImage}></video>

      <Link to="/" type="button" className="player__exit" >Exit</Link>

      <div className="player__controls">
        <div className="player__controls-row">
          <div className="player__time">
            <progress className="player__progress" value={progress} max={duration}></progress>
            <div className="player__toggler" style={{left: togglerPosition + `%`}}>Toggler</div>
          </div>
          <div className="player__time-value">{getElapsedTime(duration, progress)}</div>
        </div>

        <div className="player__controls-row">
          <button type="button" className="player__play"
            onClick={onPlayBtnClick}
          >
            {isPlaying ?
              <React.Fragment>
                <svg viewBox="0 0 14 21" width="14" height="21">
                  <use xlinkHref="#pause"></use>
                </svg>
                <span>Pause</span>
              </React.Fragment>
              :
              <React.Fragment>
                <svg viewBox="0 0 19 19" width="19" height="19">
                  <use xlinkHref="#play-s"></use>
                </svg>
                <span>Play</span>
              </React.Fragment>
            }
          </button>
          <div className="player__name">Transpotting</div>

          <button onClick={onFullscreenClick} type="button" className="player__full-screen">
            <svg viewBox="0 0 27 27" width="27" height="27">
              <use xlinkHref="#full-screen"></use>
            </svg>
            <span>Full screen</span>
          </button>
        </div>
      </div>
    </div>
  );
};

Player.propTypes = {
  currentFilmId: PropTypes.number.isRequired,
  films: PropTypes.arrayOf(filmProp),
  promoFilm: filmProp
};

const mapStateToProps = ({DATA}) => ({
  films: DATA.films,
  promoFilm: DATA.promoFilm
});

export {Player};
export default connect(mapStateToProps)(Player);
