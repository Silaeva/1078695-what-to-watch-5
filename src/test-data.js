const films = [
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 1,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 2,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 3,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 4,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 5,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 6,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 7,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 8,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 9,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 10,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 11,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 12,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  }
];

const comments = [
  {
    id: 1,
    text: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    rating: 4,
    user: `Christina`,
    userId: 19,
    date: `2020-10-09T13:38:44.769Z`
  }, {
    id: 2,
    text: `This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.`,
    rating: 9,
    user: `Sophie`,
    userId: 25,
    date: `2020-11-04T13:38:44.769Z`
  }, {
    id: 1,
    text: `Unfortunately we don't have a reliable way to tell the true ratings of a movie.`,
    rating: 7,
    user: `Jack`,
    userId: 11,
    date: `2020-10-09T13:38:44.769Z`
  }
];

const noop = () => {};

const genre = `Comedy`;

export {films, comments, noop, genre};
