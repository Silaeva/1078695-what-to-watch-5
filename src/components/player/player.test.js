import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {Player} from "./player";
import {films} from "../../test-data";

it(`Should Player render correctly`, () => {
  const tree = renderer
    .create(
        <BrowserRouter>
          <Player
            currentFilmId={1}
            films={films}
            promoFilm={films[0]}
          />
        </BrowserRouter>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
