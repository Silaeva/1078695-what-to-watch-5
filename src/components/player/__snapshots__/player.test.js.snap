// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Player render correctly 1`] = `
<div
  className="player"
>
  <video
    className="player__video"
    poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg"
    src="http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4"
  />
  <a
    className="player__exit"
    href="/"
    onClick={[Function]}
    type="button"
  >
    Exit
  </a>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max={0}
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "NaN%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        00:00
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="21"
          viewBox="0 0 14 21"
          width="14"
        >
          <use
            xlinkHref="#pause"
          />
        </svg>
        <span>
          Pause
        </span>
      </button>
      <div
        className="player__name"
      >
        Transpotting
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
